### PR TITLE
IDENTITY-3422

### PR DESCRIPTION
**Pull request by DMHP**

### CODE REVIEW

File: components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/SAMLSSOAuthenticator.java [hunk 1]
```diff
@@ -168,14 +169,17 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
                             "Defaulting to Name Identifier.");
                 }
             }
+            idpSubject = (String) request.getSession().getAttribute("username");
             if (subject == null) {
-                subject = (String) request.getSession().getAttribute("username");
+                subject=idpSubject;
```
Comment (johannnallathamby) [inline components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/SAMLSSOAuthenticator.java]: Formatting issues.. need a white space before and after the '='. Please check other places too.

File: components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/DefaultSAML2SSOManager.java [hunk 2]
```diff
@@ -78,6 +78,7 @@
 import org.opensaml.xml.validation.ValidationException;
 import org.w3c.dom.Document;
 import org.w3c.dom.Element;
+import org.wso2.carbon.base.MultitenantConstants;
```
Comment (johannnallathamby) [inline components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/DefaultSAML2SSOManager.java]: Unused import ?

File: components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/DefaultSAML2SSOManager.java [hunk 3]
```diff
@@ -359,6 +364,8 @@ private void processSSOResponse(HttpServletRequest request) throws SAMLSSOExcept
 
         // Get the subject name from the Response Object and forward it to login_action.jsp
         String subject = null;
+        String nameQualifier=null;
```
Comment (malithie) [inline components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/DefaultSAML2SSOManager.java]: Format properly

File: components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/SAMLSSOAuthenticator.java [hunk 4]
```diff
@@ -158,6 +158,7 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
                     .getSession(false).getAttribute("samlssoAttributes");
 
             String subject = null;
+            String idpSubject=null;
```
Comment (johannnallathamby) [inline components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/SAMLSSOAuthenticator.java]: Still issue remains..

File: components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/SAMLSSOAuthenticator.java [hunk 5]
```diff
@@ -168,14 +169,17 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
...
             if (subject == null) {
-                subject = (String) request.getSession().getAttribute("username");
+                subject = idpSubject;
             }
             if (subject == null) {
                 throw new SAMLSSOException("Cannot find federated User Identifier");
             }
 
             Object sessionIndexObj = request.getSession(false).getAttribute(SSOConstants.IDP_SESSION);
+            String nameQualifier = (String) request.getSession().getAttribute("nameQualifier");
```
Comment (hpmtissera) [inline components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/SAMLSSOAuthenticator.java]: better define constants for this parameters

File: components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/DefaultSAML2SSOManager.java [hunk 6]
```diff
@@ -384,11 +392,13 @@ private void processSSOResponse(HttpServletRequest request) throws SAMLSSOExcept
                 throw new SAMLSSOException("Single Logout is enabled but IdP Session ID not found in SAML Assertion");
             }
             request.getSession().setAttribute(SSOConstants.IDP_SESSION, sessionId);
+            request.getSession().setAttribute("nameQualifier", nameQualifier);
```
Comment (hpmtissera) [inline components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/DefaultSAML2SSOManager.java]: define constants

File: components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/SAMLSSOAuthenticator.java (hunk 1 above)
Comment (DMHP) [inline components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/SAMLSSOAuthenticator.java]: Formatted

File: components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/DefaultSAML2SSOManager.java (hunk 2 above)
Comment (DMHP) [inline components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/DefaultSAML2SSOManager.java]: Removed

File: components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/DefaultSAML2SSOManager.java (hunk 3 above)
Comment (DMHP) [inline components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/DefaultSAML2SSOManager.java]: Formatted

File: components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/DefaultSAML2SSOManager.java (hunk 6 above)
Comment (DMHP) [inline components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/DefaultSAML2SSOManager.java]: Fixed using https://github.com/wso2/carbon-identity/pull/935

File: components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/SAMLSSOAuthenticator.java (hunk 5 above)
Comment (DMHP) [inline components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/SAMLSSOAuthenticator.java]: Fixed using https://github.com/wso2/carbon-identity/pull/935

File: components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/SAMLSSOAuthenticator.java (hunk 4 above)
Comment (DMHP) [inline components/application-authenticators/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/SAMLSSOAuthenticator.java]: Formatted

**Pull request merged**